Compute Legendre symbol for `hash_to_curve`

davidnevadoc opened the pull request:
This is an attempt to close #73 .

Unlike in Arkworks, I believe there is no clean way to access the modulus p of a PrimeField here. I have gotten it using:
``` rust 
    let modulus: BigUint =
        BigUint::from_str_radix(F::MODULUS.strip_prefix("0x").unwrap(), 16).unwrap();
```
although in the docs it is explicitly mentioned that: 

> The encoding of the modulus is implementation-specific. Generic users of the `PrimeField` trait should treat this string as opaque.

Suggestions for alternative approaches are welcome.

huitseeker commented:
Could F::ZERO - F::ONE work (as value for MODULUS - 1)?

davidnevadoc commented:
Thanks for the suggestion @huitseeker! Updated in https://github.com/privacy-scaling-explorations/halo2curves/pull/77/commits/3b7e8b3be06657e891430df65aa32ff7824b868e

huitseeker reviewed (commented):
It would be great to see a couple of tests. One curve-independent thing to test being the relationship between `legendre` and `foo.sqrt().is_some()`.

Another thought it this is pinned on receiving ` p_minus_one`, which is systematically divided by two during the computation. Could we possibly require the field definition to include a constant for (p-1)/2?

han0110 reviewed (approved):
LGTM, also think it'd be nice to have `(p-1)/2` as constant somewhere (perhaps some private helper trait).

davidnevadoc commented:
Thanks for the reviews! I have added a couple of tests in https://github.com/privacy-scaling-explorations/halo2curves/pull/77/commits/0b40c37989dc6cd51ec66ee488608e872e650680

I agree with the comments regarding fixing `(p-1)/2`, I will experiment a bit and give it a shot in a separate PR.

davidnevadoc commented:
I have tried to address the issue of pre-computing and passing around `(p-1)/2` in a different PR https://github.com/privacy-scaling-explorations/halo2curves/pull/79. 
Let me know if we can merge this PR first and add this improvement later. Thanks everybody for the feedback! :)

davidnevadoc commented:
I have added the changes from #79 here and I've also added some benchmarks as @CPerezz requested. There is practically no change in the performance of hash to curve:
```
   Finished bench [optimized] target(s) in 23.15s
     Running benches/hash_to_curve.rs (target/release/deps/hash_to_curve-c2c106ce25b99f82)
Gnuplot not found, using plotters backend
Hash to Secp256k1       time:   [147.12 µs 147.22 µs 147.35 µs]
                        change: [-0.1760% +0.0373% +0.2060%] (p = 0.73 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

Hash to Bn256           time:   [153.73 µs 153.81 µs 153.89 µs]
                        change: [-1.5987% -1.2515% -0.9976%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe

```

han0110 commented on src/hash_to_curve.rs:
```diff
@@ -128,7 +131,7 @@ where
     let gx1 = gx1 + b;
     // 15. e1 = is_square(gx1)
-    let e1 = gx1.sqrt().is_some();
+    let e1 = !gx1.ct_quadratic_non_residue();
```
why not `let e1 = gx1.ct_quadratic_residue();`

davidnevadoc replied:
`.sqrt().is_some()` will be true when called in `F::ZERO`  I wanted to ensure that `e1` was also true in that case. However, I'm not sure that `gx1 = 0` is a valid possibility.

han0110 replied:
Ah I see, you are right, just checked the spec, `is_square` returns `true` when `gx1 = 0`

han0110 commented on src/hash_to_curve.rs:
```diff
@@ -140,7 +143,7 @@ where
     let gx2 = gx2 + b;
     // 21. e2 = is_square(gx2) AND NOT e1    # Avoid short-circuit logic ops
-    let e2 = gx2.sqrt().is_some() & (!e1);
+    let e2 = !gx2.ct_quadratic_non_residue() & (!e1);
```
ditto

han0110 reviewed (approved):
LGTM! Some nitpicks that could be ignored.
